util/src/main/java/io/kubernetes/client/leaderelection/LeaderElector.java

```diff
@@ -0,0 +1,315 @@
+      executor.submit(runnable);
+      renew(cancel);
+    } finally {
+      config.getCallbacks().onStopLeading();
+      executor.shutdown();
+    }
+  }
+
+  public void cancel() {
+    cancel.set(false);
+  }
+
+  /**
+   * GetLeader returns the identity of the last observed leader or returns the empty string if no
+   * leader has yet been observed.
+   */
+  public String getLeader() {
+    return observedRecord.getHolderIdentity();
+  }
+
+  /**
+   * IsLeader returns true if the last observed leader was this client else returns false.
+   */
+  public boolean isLeader() {
+    return config.getLock().identity().equals(observedRecord.getHolderIdentity());
+  }
+
+  /**
+   * Check will determine if the current lease is expired by more than timeout.
+   *
+   * @return true for expired;
+   */
+  public boolean check(long maxTolerableExpiredLease) {
+    if (isLeader()) {
+      // If we are more than timeout seconds after the lease duration that is past the timeout
+      // on the lease renew. Time to start reporting ourselves as unhealthy. We should have
+      // died but conditions like deadlock can prevent this. (See #70819)
+      // failed election to renew leadership on lease
+      return System.currentTimeMillis() - observedTime
+          > config.getLeaseDuration() + maxTolerableExpiredLease;
+    }
+    return false;
+  }
+
+  /**
+   * acquire loops calling tryAcquireOrRenew and returns true immediately when tryAcquireOrRenew
+   * succeeds. return false if ctx signals done.
+   */
+  private boolean acquire(AtomicBoolean cancel) {
+    String desc = config.getLock().describe();
+    logger.info("attempting to acquire leader lease {}...", desc);
+    try {
+      while (!cancel.get()) {
+        boolean success = tryAcquireOrRenew();
+        maybeReportTransition();
+        if (success) {
+          config.getLock().recordEvent("became leader");
+          logger.info("successfully acquired lease {}", desc);
+          return true;
+        } else {
+          logger.debug("failed to acuqire lease {}", desc);
+        }
+        // sleep time jittered
+        Thread.sleep(config.getRetryPeriod()
+            + (long) (random.nextDouble() * JITTER_FACTOR * config.getRetryPeriod()));
+      }
+    } catch (InterruptedException e) {
+      logger.warn("acquire thread interrupted", e);
+    }
+    return false;
+  }
+
+  /**
+   * renew loops calling tryAcquireOrRenew and returns immediately when tryAcquireOrRenew fails or
+   * ctx signals done.
+   */
+  private void renew(AtomicBoolean cancel) {
+    try {
+      while (!cancel.get()) {
+        String desc = config.getLock().describe();
+        try {
+          Callable<Boolean> callable = () -> {
+            boolean done = false;
+            while (!done && !cancel.get()) {
+              done = tryAcquireOrRenew();
+              Thread.sleep(config.getRetryPeriod());
+            }
+            return done;
+          };
+          executor.invokeAll(
```

Comment:
```suggestion
          executor.scheduleAtFixedRate(() -> {
              // ... tryAcquireOrRenew
          }, 0, getRetryPeriod().toMillis(), TimeUnit.MILLISECONDS)
```

Reply:
they are two different functions:
```java
invokeAll(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit)
scheduleAtFixedRate(Runnable command, long initialDelay, long period, TimeUnit unit)
```

Reply:
isn't it better if we retry w/ the help of a scheduled thread pool?